tsl/src/continuous_aggs/invalidation_threshold.c

```diff
@@ -0,0 +1,272 @@
+ */
+static int64
+invalidation_threshold_read(int32 hyper_id, ItemPointer tid, LOCKMODE lockmode)
+{
+	Datum threshold = DatumGetInt64(PG_INT64_MIN);
+	ScanIterator iterator;
+
+	iterator = ts_scan_iterator_create(CONTINUOUS_AGGS_INVALIDATION_THRESHOLD,
+									   lockmode,
+									   CurrentMemoryContext);
+	iterator.ctx.index = catalog_get_index(ts_catalog_get(),
+										   CONTINUOUS_AGGS_INVALIDATION_THRESHOLD,
+										   CONTINUOUS_AGGS_INVALIDATION_THRESHOLD_PKEY);
+	iterator.ctx.snapshot = GetTransactionSnapshot();
+
+	ts_scan_iterator_scan_key_init(&iterator,
+								   Anum_continuous_aggs_invalidation_threshold_pkey_hypertable_id,
+								   BTEqualStrategyNumber,
+								   F_INT4EQ,
+								   Int32GetDatum(hyper_id));
+
+	ts_scanner_foreach(&iterator)
+	{
+		TupleInfo *ti = ts_scan_iterator_tuple_info(&iterator);
+		bool isnull;
+
+		threshold =
+			slot_getattr(ti->slot, Anum_continuous_aggs_invalidation_threshold_watermark, &isnull);
+
+		if (NULL != tid)
+			ItemPointerCopy(ts_scanner_get_tuple_tid(ti), tid);
+	}
+
+	ts_scan_iterator_close(&iterator);
+
+	return DatumGetInt64(threshold);
+}
+
+/*
+ * Fetch the most recent threshold in a tuple.
+ *
+ * The tuple is referenced by the passed in TID.
+ */
+static int64
+invalidation_threshold_fetch(Relation rel, const ItemPointer tid)
+{
+	TupleTableSlot *slot;
+	bool isnull;
+	Datum datum;
+	int64 threshold;
+
+	slot = MakeSingleTupleTableSlotCompat(RelationGetDescr(rel), table_slot_callbacks(rel));
+
+	if (!table_tuple_fetch_row_version(rel, tid, GetTransactionSnapshot(), slot))
+		elog(ERROR, "could not fetch latest threshold");
+
+	datum = slot_getattr(slot, Anum_continuous_aggs_invalidation_threshold_watermark, &isnull);
+	Assert(!isnull);
+	threshold = DatumGetInt64(datum);
+
+	ExecDropSingleTupleTableSlot(slot);
+
+	return threshold;
+}
+
+/*
+ * Set a new invalidation threshold, if needed.
+ *
+ * Returns true if the new threshold was set, otherwise false.
+ *
+ * Any update to a threshold (either via an update or insert) requires
+ * completely exclusive access (including blocking readers). This is necessary
+ * to ensure we do not lose invalidations; all transactions that read the
+ * invalidation threshold must have either completed, or not yet read the
+ * value. If a weaker lock was used, an insert transaction that is concurrent
+ * with a refresh could race to read the old threshold before it was updated
+ * and invalidations written during the insert (but after the update) would be
+ * lost.
+ *
+ * Since the invalidation threshold is currently stored in a shared table, all
+ * INSERTs (that read the threshold) and refreshes of continuous aggregates
+ * will be blocked. Therefore, we try to avoid heavy locking if possible by
+ * first reading the threshold with a read lock. If the threshold needs no
+ * update, we need not grab any further locks.
+ */
+bool
+continuous_agg_invalidation_threshold_set(int32 hyper_id, int64 new_threshold)
```

Comment:
We already have code that sets thresholds and acquires the right locks on tables for  materialization in materialize.c. IMO we should be looking at integrating the refresh code that can work on arbitrary windows with existing code instead of rewriting everything from scratch. we are not taking take advantage of all the bug fixes and tests that work with the existing code.

Reply:
This isn't rewriting everything from scratch. I've studied the existing code extensively and I am trying to keep as much as possible, while at the same time applying improvements, refactoring, and cleanup. But just saying we should reuse the existing code, while changing the core functionality and semantics, is easier said than done.

Obviously, any help in making sure we maintain the lessons learned  is appreciated. But, IMO, `materialize.c` does too much and conflates a lot of different functionality (setting thresholds, materializing, etc.). It is also difficult to understand, TBH. Separating the different functionality is an improvement. Thus, I am trying clean things up and break out different functionality into separate source files. I am trying to do this without also "blowing up the house" at the same time. If I would simply reuse and change the existing code, it would have ripple effects, and even affect existing tests---something I don't want to do at this point while things are still in flux. It would cause too many changes in a single PR and would make things harder to review. So, I am trying to keep things as stable as possible, introducing one distinct change at a time.

While doing the refactoring, I am also trying to fix a number of issues, for instance, the lack of memory context use. This code also improves on the previous function to set the threshold by not always grabbing an `AccessExclusiveLock` unless necessary.

Reply:
@gayyappan I've added additional commits to integrate with the existing code and to show the history of refactoring. The PR is now a lot bigger, with more changes, but hopefully this better matches your expectations.